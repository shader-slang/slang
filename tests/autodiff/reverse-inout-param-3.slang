//TEST(compute):COMPARE_COMPUTE_EX:-slang -compute -shaderobj -output-using-type
//TEST(compute, vulkan):COMPARE_COMPUTE_EX:-vk -compute -shaderobj -output-using-type
//TEST(compute):COMPARE_COMPUTE_EX:-cpu -compute -output-using-type -shaderobj

//TEST_INPUT:ubuffer(data=[0 0 0 0 0], stride=4):out,name=outputBuffer
RWStructuredBuffer<float> outputBuffer;

[BackwardDifferentiable]
void assign(inout float rs, float v)
{
    rs = v;
}

[BackwardDifferentiable]
void f(inout float p, float x)
{
    assign(p, x);
}

[numthreads(1, 1, 1)]
void computeMain(uint3 dispatchThreadID : SV_DispatchThreadID)
{
    var x = diffPair(5.0, 0.0);
    var pp = diffPair(1.0, 3.0);
    __bwd_diff(f)(pp, x);

    outputBuffer[0] = pp.p; // should be 1, since bwd_diff does not write back new primal val.
    outputBuffer[1] = x.d; // 3
}