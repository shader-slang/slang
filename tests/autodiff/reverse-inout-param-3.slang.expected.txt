type: float
1.000000
3.000000
0.000000
0.000000
0.000000

